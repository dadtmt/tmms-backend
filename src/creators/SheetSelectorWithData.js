import R from 'ramda'
import { graphql } from 'react-apollo'
import { connect } from 'react-redux'

import { pageEditorId } from '../config'
import makeReducer from '../reducers/makeReducer'
import { GET_SHEETS_QUERY } from '../graphql/queries'
import SheetSelector from './SheetSelector'

const updatePageEditor = update => R.over(
  R.lensPath(['viewer', 'user', 'editors', 'edges']),
  R.over(
    R.lensIndex(0),
    update
  )
)

const mutationHandlers = {
  CreateSheet: (state, { createSheets }) => updatePageEditor(
    R.over(
      R.lensPath(['node', 'sheets', 'edges']),
      R.append(createSheets.changedEdge)
    )
  )(state)
}

const handlers = {
  APOLLO_MUTATION_RESULT: (state, {
    operationName,
    result: { data }
  }) => R.propOr(
    R.identity,
    operationName,
    mutationHandlers
  )(state, data)
}

export const reducer = makeReducer(handlers)

const withData = graphql(
  GET_SHEETS_QUERY,
  {
    options: {
      reducer,
      variables: {
        pageEditorId
      }
    }
  }
)

export const getSheetsFromData = R.pipe(
  R.pathOr([], ['viewer', 'user', 'editors', 'edges']),
  R.head,
  R.pathOr([], ['node', 'sheets', 'edges'])
)

const emptyCharacter = {
  name: 'new character name'
}

export const getSheetByIdFromData = id => R.pipe(
  getSheetsFromData,
  R.indexBy(R.path(['node', 'id'])),
  R.pathOr(emptyCharacter, [id, 'node'])
)

const mapStateToProps = ({ sheetSelector: { selectedSheet } }, { data }) => ({
  selectedSheet,
  sheets: getSheetsFromData(data)
})

const mapDispatchToProps = (dispatch, { data }) => ({
  selectSheet: event => dispatch({
    sheet: getSheetByIdFromData(event.target.value)(data),
    type: 'SELECT_SHEET'
  })
})

export default withData(
  connect(mapStateToProps, mapDispatchToProps)(SheetSelector)
)
