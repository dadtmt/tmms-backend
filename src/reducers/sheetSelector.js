import R from 'ramda'

import { createReducer } from './makeReducer'

const mutationHandlers = {
  CreateSheet: (state, { createSheets }) => R.assoc(
    'selectedSheet',
    R.path(['changedEdge', 'node'])(createSheets)
  )(state)
}

const handlers = {
  APOLLO_MUTATION_RESULT: (state, {
    operationName,
    result: { data }
  }) => R.propOr(
    R.identity,
    operationName,
    mutationHandlers
  )(state, data),
  SELECT_SHEET: (state, { sheet }) => R.assoc('selectedSheet', sheet)(state)
}

const initialState = {
  selectedSheet: {
    characs: [],
    gear: [],
    name: ''
  }
}

const reducer = createReducer(initialState, handlers)

export default reducer
