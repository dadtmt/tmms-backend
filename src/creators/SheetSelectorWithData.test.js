import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo'
import { createStore, combineReducers } from 'redux'

import sheetSelector from '../reducers/sheetSelector'
import SheetSelectorWithData, {
  getSheetsFromData,
  getSheetByIdFromData
} from './SheetSelectorWithData'

/*eslint-disable*/
const data = {
  'variables': {
    'pageEditorId': 'UGFnZUVkaXRvcjox'
  },
  'loading': false,
  'networkStatus': 7,
  'viewer': {
    'user': {
      'editors': {
        'edges': [
          {
            'node': {
              'id': 'UGFnZUVkaXRvcjox',
              'sheets': {
                'edges': [
                  {
                    'node': {
                      'id': 'id_1',
                      'charateristics': null,
                      'description': 'sdfsdf',
                      'image': 'dsfdf',
                      'inventory': null,
                      'isPlayer': true,
                      'name': 'Jojo le first',
                      '__typename': 'Sheets'
                    },
                    '__typename': 'SheetsEdge'
                  },
                  {
                    'node': {
                      'id': 'id_2',
                      'charateristics': null,
                      'description': null,
                      'image': null,
                      'inventory': null,
                      'isPlayer': false,
                      'name': 'toto',
                      '__typename': 'Sheets'
                    },
                    '__typename': 'SheetsEdge'
                  }
                ],
                '__typename': 'SheetsConnection'
              },
              '__typename': 'PageEditor'
            },
            '__typename': 'UserEditorsEdge'
          }
        ],
        '__typename': 'UserEditorsConnection'
      },
      '__typename': 'User'
    },
    '__typename': 'Viewer'
  }
}

/*eslint-enable*/

describe('getSheetsFromData', () => {
  it('should get sheets', () => {
    expect(getSheetsFromData(data)).toMatchSnapshot()
  })
})

describe('getSheetByIdFromData', () => {
  it('should get sheets', () => {
    expect(getSheetByIdFromData('id_2')(data)).toMatchSnapshot()
  })
})

const store = createStore(combineReducers({ sheetSelector }))
const client = new ApolloClient()

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <ApolloProvider client={client} store={store} >
      <SheetSelectorWithData />
    </ApolloProvider>,
    div
  )
})

it('renders correctly', () => {
  expect(
    renderer.create(<ApolloProvider client={client} store={store} >
      <SheetSelectorWithData />
    </ApolloProvider>).toJSON()
  ).toMatchSnapshot()
})
