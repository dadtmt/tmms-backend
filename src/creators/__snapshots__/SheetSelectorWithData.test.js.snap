exports[`getSheetByIdFromData should get sheets 1`] = `
Object {
  "__typename": "Sheets",
  "charateristics": null,
  "description": null,
  "id": "id_2",
  "image": null,
  "inventory": null,
  "isPlayer": false,
  "name": "toto",
}
`;

exports[`getSheetsFromData should get sheets 1`] = `
Array [
  Object {
    "__typename": "SheetsEdge",
    "node": Object {
      "__typename": "Sheets",
      "charateristics": null,
      "description": "sdfsdf",
      "id": "id_1",
      "image": "dsfdf",
      "inventory": null,
      "isPlayer": true,
      "name": "Jojo le first",
    },
  },
  Object {
    "__typename": "SheetsEdge",
    "node": Object {
      "__typename": "Sheets",
      "charateristics": null,
      "description": null,
      "id": "id_2",
      "image": null,
      "inventory": null,
      "isPlayer": false,
      "name": "toto",
    },
  },
]
`;

exports[`test renders correctly 1`] = `
<div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="selectCharacter">
      Select Character
    </label>
    <select
      className="form-control"
      id="selectCharacter"
      onChange={[Function]}
      placeholder="select"
      type={undefined}
      value={undefined}>
      <option
        value="">
        create a new character
      </option>
    </select>
  </div>
  <form
    onSubmit={[Function]}>
    <div
      className="checkbox"
      style={undefined}>
      <label>
        <input
          checked=""
          disabled={false}
          name="isPlayer"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value="" />
        player sheet
      </label>
    </div>
    <div
      className="form-group">
      <label
        className="control-label"
        htmlFor="name">
        Character name:
      </label>
      <input
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder={undefined}
        type="text"
        value="" />
      <span
        className="help-block" />
    </div>
    <div
      className="form-group">
      <label
        className="control-label"
        htmlFor="description">
        description:
      </label>
      <input
        className="form-control"
        id="description"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder={undefined}
        type="text"
        value="" />
      <span
        className="help-block" />
    </div>
    <div
      className="form-group">
      <label
        className="control-label"
        htmlFor={undefined}>
        image:
      </label>
      <input
        className="form-control"
        id={undefined}
        name="image"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Image url"
        type="text"
        value="" />
      <img
        className="img-rounded"
        src="" />
    </div>
    <div>
      <div
        className="form-group">
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-plus" />
        </button>
      </div>
      <ul
        className="list-group" />
    </div>
    <div>
      <div
        className="form-group">
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button">
          <span
            className="glyphicon glyphicon-plus" />
        </button>
      </div>
      <ul
        className="list-group" />
    </div>
    <div
      className="form-group">
      <button
        className="btn btn-default"
        disabled={false}
        title="save"
        type="submit">
        <span
          className="glyphicon glyphicon-save" />
      </button>
    </div>
  </form>
</div>
`;
